File: app/src/main/java/com/cere/csplayer/entity/Music.java
package com.cere.csplayer.entity;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.Nullable;

import com.litesuits.orm.db.annotation.Column;
import com.litesuits.orm.db.annotation.Default;
import com.litesuits.orm.db.annotation.Ignore;
import com.litesuits.orm.db.annotation.NotNull;
import com.litesuits.orm.db.annotation.PrimaryKey;
import com.litesuits.orm.db.annotation.Table;
import com.litesuits.orm.db.enums.AssignType;

import java.text.Collator;
import java.util.Locale;

/**
 * Created by CheRevir on 2019/10/22
 */
@Table("music")
public class Music implements Comparable<Music>, Parcelable {
    @Ignore
    private static final Collator mCollator = Collator.getInstance(Locale.CHINA);
    @PrimaryKey(AssignType.AUTO_INCREMENT)
    @Column("_id")
    private long id;
    @Column("title")
    private String title;
    @Column("artist")
    private String artist;
    @Column("album")
    private String album;
    @Column("duration")
    private int duration;
    @NotNull
    @Default("0")
    @Column("star")
    private int star;
    @Column("path")
    private String path;

    public Music() {
    }

    protected Music(Parcel parcel) {
        title = parcel.readString();
        artist = parcel.readString();
        album = parcel.readString();
        duration = parcel.readInt();
        star = parcel.readInt();
        path = parcel.readString();
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getArtist() {
        return artist;
    }

    public void setArtist(String artist) {
        this.artist = artist;
    }

    public String getAlbum() {
        return album;
    }

    public void setAlbum(String album) {
        this.album = album;
    }

    public int getDuration() {
        return duration;
    }

    public void setDuration(int duration) {
        this.duration = duration;
    }

    public int getStar() {
        return star;
    }

    public void setStar(int star) {
        this.star = star;
    }

    public String getPath() {
        return path;
    }

    public void setPath(String path) {
        this.path = path;
    }

    @Override
    public int hashCode() {
        return path.hashCode();
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (!(obj instanceof Music)) {
            return false;
        }
        return ((Music) obj).getPath().equals(this.getPath());
    }

    @Override
    public int compareTo(Music o) {
        return mCollator.compare(o.getTitle(), this.getTitle());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(title);
        dest.writeString(artist);
        dest.writeString(album);
        dest.writeInt(duration);
        dest.writeInt(star);
        dest.writeString(path);
    }

    public static final Creator<Music> CREATOR = new Creator<Music>() {
        @Override
        public Music createFromParcel(Parcel source) {
            return new Music(source);
        }

        @Override
        public Music[] newArray(int size) {
            return new Music[size];
        }
    };
}

File: app/src/main/java/com/cere/csplayer/entity/Music.kt
package com.cere.csplayer.entity

import android.net.Uri
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import android.provider.MediaStore
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import java.io.Serializable
import java.text.Collator
import java.util.*

/**
 * Created by CheRevir on 2020/9/5
 */
@Entity(tableName = "music")
data class Music(
    @PrimaryKey(autoGenerate = false) @ColumnInfo(name = "id") var id: Int,
    @ColumnInfo(name = "title") var title: String = "",
    @ColumnInfo(name = "artist") var artist: String = "",
    @ColumnInfo(name = "album") var album: String = "",
    @ColumnInfo(name = "duration") var duration: Int = 0,
    @ColumnInfo(name = "parent") var parent: String = "",
    @ColumnInfo(name = "star", defaultValue = "0") var star: Float = 0f
) : Parcelable, Comparable<Music>, Serializable {

    constructor(parcel: Parcel) : this(
        parcel.readInt(),
        parcel.readString()!!,
        parcel.readString()!!,
        parcel.readString()!!,
        parcel.readInt(),
        parcel.readString()!!,
        parcel.readFloat()
    )

    fun getData(): Uri {
        return Uri.withAppendedPath(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, id.toString())
    }

    override fun compareTo(other: Music): Int {
        return collator.compare(title, other.title)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other is Music) {
            return id == other.id
        }
        return false
    }

    override fun hashCode(): Int {
        return id
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(id)
        parcel.writeString(title)
        parcel.writeString(artist)
        parcel.writeString(album)
        parcel.writeInt(duration)
        parcel.writeString(parent)
        parcel.writeFloat(star)
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun toString(): String {
        return "Music(id=$id, title='$title', artist='$artist', album='$album', duration=$duration, star=$star, parent='$parent')"
    }

    companion object CREATOR : Creator<Music> {
        @Ignore
        val collator: Collator = Collator.getInstance(Locale.CHINA)

        override fun createFromParcel(parcel: Parcel): Music {
            return Music(parcel)
        }

        override fun newArray(size: Int): Array<Music?> {
            return arrayOfNulls(size)
        }
    }
}